File: src/main/scala/jk_5/nailed/server/scoreboard/NailedObjective.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.scoreboard

import jk_5.nailed.api.player.Player
import jk_5.nailed.api.scoreboard.{Objective, Score}
import jk_5.nailed.api.util.Checks
import jk_5.nailed.server.player.NailedPlayer
import net.minecraft.network.play.server.{S3BPacketScoreboardObjective, S3CPacketUpdateScore}

import scala.collection.mutable

/**
 * No description given
 *
 * @author jk-5
 */
case class NailedObjective(id: String, manager: NetworkedScoreboardManager) extends Objective {

  val scores = mutable.HashSet[Score]()
  val scoresByName = mutable.HashMap[String, Score]()

  var displayName: String = id

  override def setDisplayName(displayName: String){
    Checks.notNull(displayName, "displayName may not be null")
    Checks.check(displayName.length() <= 32, "displayName may not be longer than 32")

    this.displayName = displayName
    val packet = new S3BPacketScoreboardObjective
    packet.field_149343_a = this.id
    packet.field_149341_b = this.displayName
    packet.field_149342_c = 2 //Update
    this.manager.sendPacket(packet)
  }

  override def score(name: String): Score = {
    Checks.notNull(name, "name may not be null")
    scoresByName.get(name) match {
      case Some(s) => s
      case None =>
        val score = new NailedScore(this, name)
        this.scores += score
        this.scoresByName.put(name, score)
        score
    }
  }

  override def removeScore(score: Score){
    Checks.notNull(score, "score may not be null")
    if(this.scores.remove(score)){
      this.scoresByName.remove(score.name)
      val p = new S3CPacketUpdateScore
      p.name = score.name
      p.action = S3CPacketUpdateScore.Action.REMOVE
      manager.sendPacket(p)
    }
  }

  def sendData(player: Player){
    for(score <- this.scores){
      val p = new S3CPacketUpdateScore
      p.name = score.name
      p.objective = this.id
      p.value = score.value
      p.action = S3CPacketUpdateScore.Action.CHANGE
      player.asInstanceOf[NailedPlayer].sendPacket(p)
    }
  }
}

File: src/main/java/jk_5/nailed/server/scoreboard/NailedObjective.java
package jk_5.nailed.server.scoreboard;

import jk_5.nailed.api.player.Player;
import jk_5.nailed.api.scoreboard.Objective;
import jk_5.nailed.api.scoreboard.Score;
import jk_5.nailed.api.util.Checks;
import jk_5.nailed.server.player.NailedPlayer;
import net.minecraft.network.play.server.S3BPacketScoreboardObjective;
import net.minecraft.network.play.server.S3CPacketUpdateScore;
import net.minecraft.scoreboard.IScoreObjectiveCriteria;

import javax.annotation.Nonnull;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class NailedObjective implements Objective {

    private final String id;
    private final NetworkedScoreboardManager manager;
    private final Set<Score> scores = new HashSet<Score>();
    private final Map<String, Score> scoresByName = new HashMap<String, Score>();

    private String displayName;

    public NailedObjective(String id, NetworkedScoreboardManager manager) {
        this.id = id;
        this.manager = manager;
        this.displayName = id;
    }

    @Nonnull
    @Override
    public String getId() {
        return this.id;
    }

    @Nonnull
    @Override
    public String getDisplayName() {
        return this.displayName;
    }

    @Override
    public void setDisplayName(@Nonnull String displayName) {
        Checks.notNull(displayName, "displayName may not be null");
        Checks.check(displayName.length() <= 32, "displayName may not be longer than 32");

        this.displayName = displayName;
        S3BPacketScoreboardObjective packet = new S3BPacketScoreboardObjective();
        packet.objectiveName = this.id;
        packet.objectiveValue = this.displayName;
        packet.type = IScoreObjectiveCriteria.EnumRenderType.INTEGER; //TODO: config option
        packet.field_149342_c = 2; //Update
        this.manager.sendPacket(packet);
    }

    @Nonnull
    @Override
    public Score getScore(@Nonnull String name) {
        Checks.notNull(name, "name may not be null");
        Score score = this.scoresByName.get(name);
        if(score == null){
            score = new NailedScore(this, name);
            this.scores.add(score);
            this.scoresByName.put(name, score);
        }
        return score;
    }

    @Override
    public void removeScore(@Nonnull Score score) {
        Checks.notNull(score, "score may not be null");
        if(this.scores.remove(score)){
            this.scoresByName.remove(score.getName());
            S3CPacketUpdateScore p = new S3CPacketUpdateScore();
            p.name = score.getName();
            p.action = S3CPacketUpdateScore.Action.REMOVE;
            manager.sendPacket(p);
        }
    }

    public void sendData(Player player){
        for(Score score : this.scores){
            S3CPacketUpdateScore p = new S3CPacketUpdateScore();
            p.name = score.getName();
            p.objective = this.id;
            p.value = score.getValue();
            p.action = S3CPacketUpdateScore.Action.CHANGE;
            ((NailedPlayer) player).sendPacket(p);
        }
    }
}
